File: golden-paths/ruby/rails/app/controllers/golden_controller.rb
# frozen_string_literal: true

# The whole golden-path contract in one controller: /health /ready /hello /metrics.
class GoldenController < ApplicationController
  SERVICE_NAME = "golden-ruby-rails"

  def health
    render json: { status: "ok" }
  end

  def ready
    render json: { status: "ready" }
  end

  def hello
    render json: { service: SERVICE_NAME, message: "hello, weyland" }
  end

  # Prometheus text exposition, hand-rolled to stay dependency-free — a real service swaps in the
  # `prometheus-client` gem with a proper registry and process metrics.
  def metrics
    body = "# HELP golden_hello_requests_total Calls to /hello\n" \
           "# TYPE golden_hello_requests_total counter\n" \
           "golden_hello_requests_total 0\n"
    render plain: body, content_type: "text/plain; version=0.0.4"
  end
end

File: golden-paths/elixir/phoenix/lib/golden_phoenix_web/controllers/golden_controller.ex
defmodule GoldenPhoenixWeb.GoldenController do
  @moduledoc "The whole golden-path contract in one controller: /health /ready /hello /metrics."
  use GoldenPhoenixWeb, :controller

  @service "golden-elixir-phoenix"

  def health(conn, _params), do: json(conn, %{status: "ok"})
  def ready(conn, _params), do: json(conn, %{status: "ready"})

  def hello(conn, _params) do
    json(conn, %{service: @service, message: "hello, weyland"})
  end

  # Prometheus text exposition, hand-rolled to stay dependency-free.
  def metrics(conn, _params) do
    body =
      "# HELP golden_hello_requests_total Calls to /hello\n" <>
        "# TYPE golden_hello_requests_total counter\n" <>
        "golden_hello_requests_total 0\n"

    conn
    |> put_resp_content_type("text/plain; version=0.0.4")
    |> send_resp(200, body)
  end
end
